machine_learning/neural_network.cpp

```diff
@@ -0,0 +1,719 @@
+                            } 
+                            x_data = pop_back(x_data); // Remove label from x_data
+                        }
+                        else {
+                            y_data.resize(this -> layers.back().neurons);
+                            // If task is classification
+                            if(y_data.size() > 1) {
+                                y_data[x_data[x_data.size() - 1]] = 1;
+                            }
+                            // If task is regrssion (of single value)
+                            else {
+                                y_data[0] = x_data[x_data.size() - 1];
+                            } 
+                            x_data = pop_front(x_data); // Remove label from x_data
+                        }
+                        // Push collected X_data and y_data in X and Y
+                        X.push_back({x_data});
+                        Y.push_back({y_data});
+                    }
+                    in_file.close();
+                    // Normalize training data if flag is set
+                    if(normalize) {
+                        // Scale data between 0 and 1 using min-max scaler
+                        X = minmax_scaler(X, 0.01, 1.0);
+                    }
+                    return make_pair(X, Y); // Return pair of X and Y
+                }
+
+                /**
+                 * Function to get prediction of model on single sample.
+                 * @param X X data
+                 * @return returns predictions as vector
+                 */  
+                std::vector<std::valarray <double>> 
+                single_predict (const std::vector<std::valarray <double>> &X) {
+                    // Get activations of all layers
+                    auto activations = this -> __detailed_single_prediction(X);
+                    // Return activations of last layer (actual predicted values)
+                    return activations.back();
+                }
+
+                /**
+                 * Function to get prediction of model on batch
+                 * @param X X data
+                 * @return returns predicted values as vector
+                 */  
+                std::vector < std::vector <std::valarray<double>>>
+                batch_predict (const std::vector <std::vector <std::valarray <double>>> &X) {
+                    // Store predicted values
+                    std::vector < std::vector <std::valarray<double>>> predicted_batch(X.size());
+                    for(size_t i = 0; i < X.size(); i++) { // For every sample
+                        // Push predicted values
+                        predicted_batch[i] = this -> single_predict(X[i]);
+                    }
+                    return predicted_batch; // Return predicted values
+                }
+
+                /**
+                 * Function to fit model on supplied data
+                 * @param X X data
```

Comment:
Would it be better to describe `X` as `array of feature vectors` and `Y` as `array of known/expected classifications`?